src/Features/LanguageServer/Protocol/CustomProtocol/FindUsagesLSPContext.cs

```diff
@@ -0,0 +1,211 @@
+        private readonly int _position;
+        private readonly IMetadataAsSourceFileService _metadataAsSourceFileService;
+
+        public override CancellationToken CancellationToken { get; }
+
+        public FindUsagesLSPContext(
+            IProgress<object[]> progress,
+            Document document,
+            int position,
+            IMetadataAsSourceFileService metadataAsSourceFileService,
+            CancellationToken cancellationToken)
+        {
+            _progress = progress;
+            _document = document;
+            _position = position;
+            _metadataAsSourceFileService = metadataAsSourceFileService;
+
+            CancellationToken = cancellationToken;
+        }
+
+        public override Task OnCompletedAsync()
+        {
+            var results = Array.Empty<VSReferenceItem>();
+            lock (_gate)
+            {
+                if (_resultsChunk.IsEmpty())
+                {
+                    return Task.CompletedTask;
+                }
+
+                results = _resultsChunk.ToArray();
+                _resultsChunk.Clear();
+            }
+
+            _progress.Report(results);
+            return Task.CompletedTask;
+        }
+
+        public async override Task OnDefinitionFoundAsync(DefinitionItem definition)
+        {
+            if (!_definitionToId.ContainsKey(definition))
+            {
+                // Assinging a new id to the definition
+                _definitionToId.Add(definition, _id);
+
+                // Creating a new VSReferenceItem for the definition
+                var definitionItem = await GenerateVSReferenceItemAsync(
+                    _id, definitionId: _id, _document, _position, definition.SourceSpans.FirstOrDefault(),
+                    definition.DisplayableProperties, _metadataAsSourceFileService, definition.GetClassifiedText(),
+                    symbolUsageInfo: null, CancellationToken).ConfigureAwait(false);
+
+                if (definitionItem.Location != null)
+                {
+                    AddAndReportResultsIfAtMax(definitionItem);
+                    _id++;
+                }
+            }
+        }
+
+        public async override Task OnReferenceFoundAsync(SourceReferenceItem reference)
+        {
+            if (_definitionToId.TryGetValue(reference.Definition, out var definitionId))
+            {
+                // Creating a new VSReferenceItem for the reference
+                var referenceItem = await GenerateVSReferenceItemAsync(
+                    _id, definitionId, _document, _position, reference.SourceSpan,
+                    reference.AdditionalProperties, _metadataAsSourceFileService, definitionText: null,
+                    reference.SymbolUsageInfo, CancellationToken).ConfigureAwait(false);
+
+                if (referenceItem.Location != null)
+                {
+                    AddAndReportResultsIfAtMax(referenceItem);
+                    _id++;
```

Comment:
btw, this is not threadsafe.  FindRefs is multi-threaded and may have many clients calling OnReferenceFoundAsync concurrently.